### PR TITLE
8 add support for cookie authentication

### LINKED ISSUE #8: Add support for Cookie authentication
1. Create new generator/command to opt-in for cookie authentication named pkg-cookie-auth. This will:
- Install needed libraries (such as iron-session)
- Flag "cookie-auth" package true in the GeNYG config file
2. Create new generator/command to create a new user type/role (such as user, admin, editor, publisher, etc) named cookie-auth-role (yo g-next:cookie-auth-role). So if I run `yo g-next:cookie-auth` and as role I input "admin", this will create some structure similar to this:
```ts
export const adminSessionOptions: IronSessionOptions = {
  password: process.env.ADMIN_SECRET_COOKIE_PASSWORD as string,
  cookieName: "name-of-the-project-cookie-auth-admin",
  cookieOptions: {
    secure: process.env.NODE_ENV === "production",
  },
};
```
We can do it for example by modifying .genyg.json file with a new entry: `"cookie-auth-roles": ["admin", "user"]`
3. Modify `yo g-next:api` command in order to support cookie auth templates. So after path/route and method parameters prompt, the generator should check if cookie-auth is used, and if so it should prompt if we want to use authentication or not with a list of options such as: "public API", "cookie protected API". If cookie is selected then we want to list of roles (from .genyg.json cookie-auth-roles file), and create index, handler, etc. files accordingly.

### PR DESCRIPTION
**Pull request by BoulosMansour**

### CODE REVIEW

**Review by antoniogiordano — CHANGES_REQUESTED** [review]
Hi @BoulosMansour , the commands are still buggy. Try to pull my modifications, then you can run (inside your generator project folder!!):
```sh
npm uninstall generator-g-next -g
npm install -g
```
This will install your local copy of the generator globally.
Now you can create a new GeNYG project and test your commands ;-)

**Review by antoniogiordano — CHANGES_REQUESTED** [review]
Good job @BoulosMansour , as next improvements we should:
1. Rename the server models from "Admin" or "User" to "AdminSession" and "UserSession" both for the folder and the class itself
2. Modify also the handler of the APIs in order to handle the req.session object, with this snippet:
```ts
export default async function handler(
  req: DeleteArticlesByArticleIdApi.Request,
  res: NextApiResponse<DeleteArticlesByArticleIdApi.EndpointResponse>,
  originalReq: NextApiRequest,
) {
  try {
    const { validationResult, queryStringParameters } = req;

    if (!validationResult.isValid) {
      return ResponseHandler.json<ErrorResponse>(
        res,
        { message: validationResult.message! },
        StatusCodes.BadRequest,
      );
    }

    if (!originalReq.session.admin) {
      return ResponseHandler.json<ErrorResponse>(
        res,
        {},
        StatusCodes.Unauthorized,
      );
    }

...

```
3. Then we want to do a similar things with pages. In fact also pages could be protected with cookie, but only if they are ServerSide rendered. You can check the pages command, users can select ServerSideRendering, in that case we need to prompt for cookie auth and role. Please check hoe I modified the api generator to retrieve information in answers object from prompt.

**Comment by antoniogiordano** [comment]
@BoulosMansour This is a snippet for cookie protected pages:
```ts
import { withIronSessionSsr } from "iron-session/next";

...

export const getServerSideProps = withIronSessionSsr(
  async function getServerSideProps({
    params: { slugs },
    req: { session, cookies, headers },
  }: GetServerSidePropsContext<{ slugs: string[] }>): Promise<
    GetServerSidePropsResult<GenericPageProps>
  > {

...

```

File: generators/cookie-auth-role/index.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    extendConfigFile(this, {
+      cookieRoles,
+    });
+
+    //Index.tsx session file
+    this.fs.write(
+      this.destinationPath("./src/lib/session/index.tsx"),
+      generateLibSessionFile({
+        cookieRoles,
+        projectName: require(this.destinationPath("./package-lock.json")).name,
```
Comment (antoniogiordano) [inline generators/cookie-auth-role/index.js]:
```suggestion
        projectName: require(this.destinationPath("./package.json")).name,
```
why package lock?

**Review by antoniogiordano — CHANGES_REQUESTED** [review]
Hi @BoulosMansour , if you create a role with a whitespace, like "master operator" and you create a page with SSR, the page is not working properly

File: generators/cookie-auth-role/index.js (hunk 1 above)
Comment (BoulosMansour) [reply generators/cookie-auth-role/index.js]: when I am creating a new project I have no parameter called "name" in package.json 
Maybe I am missing something

**Review by antoniogiordano — CHANGES_REQUESTED** [review]
You can specify the name manually in the package json.

Anyway, still not good for pages, try always locally before please:

Screenshot 2023-02-24 at 14 21 00

**Review by antoniogiordano — APPROVED** [review]

**Pull request merged**